[macOS] Dragging and dropping photos from PHPickerViewController results in 0 byte files

pxlcoder opened the pull request:
#### 758a3d5efac98095352c9755ed1894f9b1d1f638
<pre>
[macOS] Dragging and dropping photos from PHPickerViewController results in 0 byte files
<a href="https://bugs.webkit.org/show_bug.cgi?id=291245">https://bugs.webkit.org/show_bug.cgi?id=291245</a>
<a href="https://rdar.apple.com/145098086">rdar://145098086</a>

Reviewed by Wenson Hsieh.

Photos dragged out of `PHPickerViewController` are backed by a file promise URL.
However, the file promise URLs do not use `NSFilePromiseReceiver` and
`NSFilesPromisePboardType` is not written to the pasteboard. Consequently, file
coordination must be used to obtain the actual file.

Fix by updating the drag and drop for files to use file coordination. Note that
file coordination works regardless of whether or not the file is backed by a
promise.

No new tests, since after consultation with FileProvider experts it is not
easily possible to create a file URL that *requires* coordination. Existing
tests cover the existing support for dragging and dropping files.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::handleLegacyFilesPromisePasteboard):

Rename method to make it more clear it handles `NSFilesPromisePboardType`.

(WebKit::handleLegacyFilesPasteboard):

Use `NSFileCoordinator` to fulfill file promises. `prepareForReadingItemsAtURLs:`
and `coordinateReadingItemAtURL:` are synchronous, so they are performed on a
background thread.

(WebKit::WebViewImpl::performDragOperation):

Canonical link: <a href="https://commits.webkit.org/293453@main">https://commits.webkit.org/293453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97ff4fc52997de479d2ec531ffb7e548a2acfbc5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98919 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104045 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49510 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27006 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75303 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32432 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89332 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55664 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14113 "Found 2 new test failures: imported/w3c/web-platform-tests/beacon/beacon-cors.https.window.html imported/w3c/web-platform-tests/css/css-position/sticky/position-sticky-fixed-ancestor-iframe.html (failure)") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48888 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84050 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106414 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26016 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18962 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84266 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26391 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83766 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28419 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6094 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19738 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16094 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25969 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31155 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29109 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27363 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4249,6 +4249,50 @@ static bool handleLegacyFilesPasteboard(id<NSDraggingInfo> draggingInfo, Box<Web
+    for (NSString *file in files.get())
+        [originalFileURLs addObject:adoptNS([[NSURL alloc] initFileURLWithPath:file]).get()];
+
+    auto task = makeBlockPtr([protectedPage = Ref { page }, originalFileURLs, dragData, pasteboardName = pasteboardName.isolatedCopy()]() mutable {
```
```suggestion
    auto task = makeBlockPtr([protectedPage = Ref { page }, originalFileURLs, dragData, pasteboardName = pasteboardName.isolatedCopy()] mutable {
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4249,6 +4249,50 @@ static bool handleLegacyFilesPasteboard(id<NSDraggingInfo> draggingInfo, Box<Web
+                RELEASE_LOG_ERROR_IF(error, DragAndDrop, "Failed to coordinate reading file: %@.", error.localizedDescription);
+            }
+
+            RunLoop::protectedMain()->dispatch([protectedPage = WTFMove(protectedPage), fileNames, dragData, pasteboardName, completionHandler = makeBlockPtr(completionHandler)] () mutable {
```
```suggestion
            RunLoop::protectedMain()->dispatch([protectedPage = WTFMove(protectedPage), fileNames, dragData, pasteboardName, completionHandler = makeBlockPtr(completionHandler)] mutable {
```

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4249,6 +4249,50 @@ static bool handleLegacyFilesPasteboard(id<NSDraggingInfo> draggingInfo, Box<Web
+
+            for (NSURL *originalFileURL in originalFileURLs.get()) {
+                NSError *error = nil;
+                [coordinator coordinateReadingItemAtURL:originalFileURL options:NSFileCoordinatorReadingWithoutChanges error:&error byAccessor:[fileNames](NSURL *newURL) {
+                    fileNames->append(newURL.path);
+                }];
```
Oh, took me a moment to realize that `-coordinateReadingItemAtURL:…:` was sync — I was wondering what guaranteed that `fileNames` would be populated by the time it was moved into `performDragWithLegacyFiles`

whsieh reviewed (commented):
Seems reasonable to me! Will hold off on the r+ until it's out of Draft state

pxlcoder committed:
[macOS] Dragging and dropping photos from PHPickerViewController results in 0 byte files
https://bugs.webkit.org/show_bug.cgi?id=291245
rdar://145098086

Reviewed by Wenson Hsieh.

Photos dragged out of `PHPickerViewController` are backed by a file promise URL.
However, the file promise URLs do not use `NSFilePromiseReceiver` and
`NSFilesPromisePboardType` is not written to the pasteboard. Consequently, file
coordination must be used to obtain the actual file.

Fix by updating the drag and drop for files to use file coordination. Note that
file coordination works regardless of whether or not the file is backed by a
promise.

No new tests, since after consultation with FileProvider experts it is not
easily possible to create a file URL that *requires* coordination. Existing
tests cover the existing support for dragging and dropping files.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::handleLegacyFilesPromisePasteboard):

Rename method to make it more clear it handles `NSFilesPromisePboardType`.

(WebKit::handleLegacyFilesPasteboard):

Use `NSFileCoordinator` to fulfill file promises. `prepareForReadingItemsAtURLs:`
and `coordinateReadingItemAtURL:` are synchronous, so they are performed on a
background thread.

(WebKit::WebViewImpl::performDragOperation):

Canonical link: https://commits.webkit.org/293453@main
